src/Mono.Android/Xamarin.Android.Net/AndroidMessageHandler.cs

```diff
@@ -157,13 +157,13 @@ public CookieContainer CookieContainer
 
 		public bool CheckCertificateRevocationList { get; set; } = false;
 
-		X509TrustManagerWithValidationCallback.Helper? _callbackTrustManagerHelper = null;
+		ServerCertificateCustomValidator? _serverCertificateCustomValidator = null;
 
 		public Func<HttpRequestMessage, X509Certificate2?, X509Chain?, SslPolicyErrors, bool>? ServerCertificateCustomValidationCallback
 		{
-			get => _callbackTrustManagerHelper?.Callback;
+			get => _serverCertificateCustomValidator?.Callback;
 			set {
-				_callbackTrustManagerHelper = value != null ? new X509TrustManagerWithValidationCallback.Helper (value) : null;
+				_serverCertificateCustomValidator = value != null ? new ServerCertificateCustomValidator (value) : null;
```

Comment:
I'm uncertain about the *structure* of this change.  Every time the `AndroidMessageHandler.ServerCertificateCustomValidationCallback` property is set, we create a new `ServerCertificateCustomValidator` instance, which in turn creates a new `AlwaysAcceptingHostnameVerifier` instance, which is a "bridged" instance (GC overheads).  

Is this necessary?

It feels like we'd be more "GC friendly" if we instead did:

```csharp
if (value == null) {
    _serverCertificateCustomValidator = null;
} else {
    _serverCertificateCustomValidator ??= new ServerCertificateCustomValidator (value);
    _serverCertificateCustomValidator.Callback = value;
}
```

Reply:
A possible concern here is thread-safety.  I can't find any documented thread-safety requirements on e.g. [`SocketsHttpHandler`](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler?view=net-6.0), though [`HttpMessageHandler`](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpmessagehandler?view=net-6.0) *does* imply that thread-safety is required.

If we thus assume that we need to be thread-safe…. Assigning to `_serverCertificateCustomValidator.Callback` is thread-safe.  What I'm less certain about is the *interaction* between `ServerCertificateCustomValidationCallback` and `SetupSSL()`, as `SetupSSL()` calls `_serverCertificateCustomValidator.ReplaceX509TrustManager()`, so *which* Callback is used could get weird/confusing.

We thus may need to *not* do `_serverCertificateCustomValidator.Callback = value`.

Reply:
Related meta-question: what are the semantics in a multi-threaded environment when intermixing `.Send()` and assignments to `.ServerCertificateCustomValidationCallback`?  Is this a Don't Do That™ scenario?

Reply:
Thanks for the suggestion. I'll update the PR to reduce allocations and make it thread-safe.

Customers can use it anyway they like, but the typical use case should be to set the callback once and then keep it the same for the lifetime of the handler. I don't think it makes too much sense to change the callbacks for different requests.

In general, there are many different properties with public setters that affect the request and that can change independently of calls to `SendAsync` (for example `PreAuthenticate`, `UseProxy`, `Proxy`, `Credentials`, `TrustedCerts`). I think the thread-safety is meant only to parallel calls to `SendAsync` (e.g., redirect state for each parallel request should be independent). IMO this is a Don't Do That™ scenario.

Reply:
I'm not sure we want to cache the instance of the custom validator. I can imagine a scenario where the validation callback is changed between different URLs (e.g. accessing a local server with a self-signed certificate and the proxying results of the query to a server on the Internet with a valid cert etc) while keeping the same instance of the handler around. With that, I think the most common scenario is where the client handler is a "one short" affair, where the app either makes requests to the same server repeatedly (keeping the client handler around) or where it makes a single request to somewhere. In both scenarios caching won't improve neither memory use nor performance. Caching means we assume `value` passed to the property doesn't change - it feels like an unnecessary restriction.

Reply:
@grendello I'm not sure if I understand your comment correctly. The custom validator class doesn't have any internal state except for the callback. Replacing the callback on an existing instance and creating a new instance with the new callbacks are functionally identical. The reason why we create an instance of the class, and we don't use just static methods to set up trust managers and the hostname verifier, is to make the code linker friendly.

Reply:
@simonrozsival it's possible I'm overthinking this. I'm thinking about a scenario when a single instance of the handler is reused for different servers and uses different callbacks and whether sharing a single instance of the verifier (with a different callback) would/could potentially break something. With callbeck being the only state of the verifier, I guess my worries are unfounded, just can't get rid of that gnawing feeling that something's off :)